File: src/services/nssai-availability.ts
import { getCollection } from '../db/mongodb';
import {
  NssaiAvailabilitySubscription,
  NssaiAvailabilitySubscriptionCreateRequest,
  NssaiAvailabilitySubscriptionUpdateRequest,
  NssaiAvailabilityNotification,
  AuthorizedNssaiAvailabilityData,
  RestrictedSnssai,
  RestrictionType
} from '../types/nnssf-nssaiavailability-types';
import { SliceConfiguration } from '../types/db-types';
import { Snssai, Tai, PlmnId } from '../types/common-types';
import { createId } from '@paralleldrive/cuid2';
import axios from 'axios';

const snssaiMatches = (s1: Snssai, s2: Snssai): boolean => {
  return s1.sst === s2.sst && s1.sd === s2.sd;
};

const plmnMatches = (p1: PlmnId, p2: PlmnId): boolean => {
  return p1.mcc === p2.mcc && p1.mnc === p2.mnc;
};

const taiMatches = (t1: Tai, t2: Tai): boolean => {
  return plmnMatches(t1.plmnId, t2.plmnId) && t1.tac === t2.tac;
};

export const createNssaiAvailabilitySubscription = async (
  request: NssaiAvailabilitySubscriptionCreateRequest
): Promise<NssaiAvailabilitySubscription> => {
  const subscriptionsCollection = getCollection<NssaiAvailabilitySubscription>('nssai_availability_subscriptions');

  const subscriptionId = createId();
  const now = new Date();

  const subscription: NssaiAvailabilitySubscription = {
    subscriptionId,
    nfInstanceId: request.nfInstanceId,
    subscriptionData: request.subscriptionData,
    notificationUri: request.notificationUri,
    supportedFeatures: request.supportedFeatures,
    expiryTime: request.expiryTime,
    createdAt: now,
    updatedAt: now
  };

  await subscriptionsCollection.insertOne(subscription);

  return subscription;
};

export const getNssaiAvailabilitySubscription = async (
  subscriptionId: string
): Promise<NssaiAvailabilitySubscription | null> => {
  const subscriptionsCollection = getCollection<NssaiAvailabilitySubscription>('nssai_availability_subscriptions');

  const subscription = await subscriptionsCollection.findOne({ subscriptionId });

  return subscription;
};

export const updateNssaiAvailabilitySubscription = async (
  subscriptionId: string,
  update: NssaiAvailabilitySubscriptionUpdateRequest
): Promise<NssaiAvailabilitySubscription | null> => {
  const subscriptionsCollection = getCollection<NssaiAvailabilitySubscription>('nssai_availability_subscriptions');

  const updateDoc = {
    $set: {
      subscriptionData: update.subscriptionData,
      supportedFeatures: update.supportedFeatures,
      expiryTime: update.expiryTime,
      updatedAt: new Date()
    }
  };

  const result = await subscriptionsCollection.findOneAndUpdate(
    { subscriptionId },
    updateDoc,
    { returnDocument: 'after' }
  );

  return result || null;
};

export const deleteNssaiAvailabilitySubscription = async (
  subscriptionId: string
): Promise<boolean> => {
  const subscriptionsCollection = getCollection<NssaiAvailabilitySubscription>('nssai_availability_subscriptions');

  const result = await subscriptionsCollection.deleteOne({ subscriptionId });

  return result.deletedCount > 0;
};

export const getAuthorizedNssaiAvailabilityData = async (
  tai: Tai,
  supportedSnssaiList?: Snssai[]
): Promise<AuthorizedNssaiAvailabilityData> => {
  const slicesCollection = getCollection<SliceConfiguration>('slices');

  const slices = await slicesCollection.find({
    'plmnId.mcc': tai.plmnId.mcc,
    'plmnId.mnc': tai.plmnId.mnc
  }).toArray();

  const availableSnssais: Snssai[] = [];
  const restrictedSnssais: RestrictedSnssai[] = [];

  const snssaisToCheck = supportedSnssaiList || slices.map(s => s.snssai);

  for (const snssai of snssaisToCheck) {
    const slice = slices.find(s => snssaiMatches(s.snssai, snssai));

    if (!slice) {
      restrictedSnssais.push({
        snssai,
        restrictionType: RestrictionType.NOT_ALLOWED
      });
      continue;
    }

    if (slice.taiList && slice.taiList.length > 0) {
      const isAvailableInTai = slice.taiList.some(sliceTai => taiMatches(sliceTai, tai));

      if (!isAvailableInTai) {
        restrictedSnssais.push({
          snssai,
          restrictionType: RestrictionType.RESTRICTED_IN_TAI
        });
        continue;
      }
    }

    availableSnssais.push(snssai);
  }

  return {
    tai,
    supportedSnssaiList: availableSnssais,
    restrictedSnssaiList: restrictedSnssais.length > 0 ? restrictedSnssais : undefined
  };
};

export const notifyNssaiAvailabilityChange = async (
  tai: Tai,
  supportedSnssaiList?: Snssai[]
): Promise<void> => {
  const subscriptionsCollection = getCollection<NssaiAvailabilitySubscription>('nssai_availability_subscriptions');

  const subscriptions = await subscriptionsCollection.find({
    'subscriptionData.tai.plmnId.mcc': tai.plmnId.mcc,
    'subscriptionData.tai.plmnId.mnc': tai.plmnId.mnc,
    'subscriptionData.tai.tac': tai.tac
  }).toArray();

  const authorizedData = await getAuthorizedNssaiAvailabilityData(tai, supportedSnssaiList);

  for (const subscription of subscriptions) {
    const notification: NssaiAvailabilityNotification = {
      subscriptionId: subscription.subscriptionId,
      authorizedNssaiAvailabilityData: [authorizedData]
    };

    try {
      await axios.post(subscription.notificationUri, notification, {
        headers: {
          'Content-Type': 'application/json'
        },
        timeout: 5000
      });
    } catch (error) {
      console.error(`Failed to send NSSAI availability notification to ${subscription.notificationUri}:`, error);
    }
  }
};

File: src/services/nssai_availability.rs
use chrono::Utc;
use mongodb::Database;
use bson::doc;
use uuid::Uuid;

use crate::errors::{AppError, AppResult};
use crate::types::db::NssaiAvailabilitySubscription;
use crate::types::nssaiavailability::{
    NssfEventSubscriptionCreateData, NssfEventSubscriptionCreatedData, PatchItem, PatchOp,
};

pub async fn create_subscription(
    data: NssfEventSubscriptionCreateData,
    db: &Database,
) -> AppResult<NssfEventSubscriptionCreatedData> {
    let collection = db.collection::<NssaiAvailabilitySubscription>(
        "nssai_availability_subscriptions",
    );

    let subscription_id = Uuid::new_v4().to_string();
    let now = Utc::now();

    let subscription = NssaiAvailabilitySubscription {
        subscription_id: subscription_id.clone(),
        nf_instance_id: data.amf_id.clone().unwrap_or_default(),
        subscription_data: crate::types::db::NssaiAvailabilitySubscriptionData {
            tai: data.tai_list.as_ref().and_then(|l| l.first()).cloned().unwrap_or(
                crate::types::common::Tai {
                    plmn_id: crate::types::common::PlmnId {
                        mcc: "000".to_string(),
                        mnc: "00".to_string(),
                    },
                    tac: "0000".to_string(),
                },
            ),
            supported_snssai_list: None,
        },
        notification_uri: data.nf_nssai_availability_uri.clone(),
        supported_features: data.supported_features.clone(),
        expiry_time: data.expiry.clone(),
        created_at: now,
        updated_at: now,
    };

    collection.insert_one(&subscription).await.map_err(|e| {
        AppError::InternalServerError(format!("Failed to create subscription: {}", e))
    })?;

    Ok(NssfEventSubscriptionCreatedData {
        subscription_id,
        expiry: data.expiry,
        authorized_nssai_availability_data: None,
        supported_features: data.supported_features,
    })
}

pub async fn get_subscription(
    subscription_id: &str,
    db: &Database,
) -> AppResult<NssaiAvailabilitySubscription> {
    let collection = db.collection::<NssaiAvailabilitySubscription>(
        "nssai_availability_subscriptions",
    );

    let subscription = collection
        .find_one(doc! { "subscriptionId": subscription_id })
        .await
        .map_err(|e| {
            AppError::InternalServerError(format!("Failed to query subscription: {}", e))
        })?;

    match subscription {
        Some(s) => Ok(s),
        None => Err(AppError::NotFound("Subscription not found".to_string())),
    }
}

pub async fn update_subscription(
    subscription_id: &str,
    patches: Vec<PatchItem>,
    db: &Database,
) -> AppResult<NssfEventSubscriptionCreatedData> {
    let collection = db.collection::<NssaiAvailabilitySubscription>(
        "nssai_availability_subscriptions",
    );

    let existing = collection
        .find_one(doc! { "subscriptionId": subscription_id })
        .await
        .map_err(|e| {
            AppError::InternalServerError(format!("Failed to query subscription: {}", e))
        })?;

    let existing = match existing {
        Some(s) => s,
        None => return Err(AppError::NotFound("Subscription not found".to_string())),
    };

    let mut update_doc = doc! {};

    for patch in &patches {
        let field = patch.path.trim_start_matches('/').replace('/', ".");

        match patch.op {
            PatchOp::Replace | PatchOp::Add => {
                if let Some(value) = &patch.value {
                    let bson_value = bson::to_bson(value).map_err(|e| {
                        AppError::BadRequest(format!("Invalid patch value: {}", e))
                    })?;
                    update_doc.insert(field, bson_value);
                }
            }
            PatchOp::Remove => {
                update_doc.insert(field, bson::Bson::Null);
            }
            _ => {
                return Err(AppError::BadRequest(format!(
                    "Unsupported patch operation: {:?}",
                    patch.op
                )));
            }
        }
    }

    update_doc.insert("updatedAt", bson::to_bson(&Utc::now()).unwrap());

    collection
        .update_one(
            doc! { "subscriptionId": subscription_id },
            doc! { "$set": update_doc },
        )
        .await
        .map_err(|e| {
            AppError::InternalServerError(format!("Failed to update subscription: {}", e))
        })?;

    Ok(NssfEventSubscriptionCreatedData {
        subscription_id: subscription_id.to_string(),
        expiry: existing.expiry_time,
        authorized_nssai_availability_data: None,
        supported_features: existing.supported_features,
    })
}

pub async fn delete_subscription(
    subscription_id: &str,
    db: &Database,
) -> AppResult<()> {
    let collection = db.collection::<NssaiAvailabilitySubscription>(
        "nssai_availability_subscriptions",
    );

    let result = collection
        .delete_one(doc! { "subscriptionId": subscription_id })
        .await
        .map_err(|e| {
            AppError::InternalServerError(format!("Failed to delete subscription: {}", e))
        })?;

    if result.deleted_count == 0 {
        return Err(AppError::NotFound("Subscription not found".to_string()));
    }

    Ok(())
}
